DEVELOP.md

```diff
@@ -0,0 +1,252 @@
+```
+After this, you should add kubebuilder and kustomize to your PATH environment:
+```
+export PATH=${PATH}:/usr/local/kubebuilder/bin
+```
+
+start docker service, if you are using centos, you can use the following command to start
+```
+service docker start
+```
+Then we can test the toolchain
+```
+make
+```
+It should work. But it's not enough yet, we need to set up a local Kubernetes cluster, which needs [kind](https://github.com/kubernetes-sigs/kind)
+```
+curl -Lo ./kind "https://github.com/kubernetes-sigs/kind/releases/download/v0.7.0/kind-$(uname)-amd64"
+chmod +x ./kind
+mv kind /usr/local/bin/
+```
+In addition to kind, you also need helm
+```
+curl -Lo helm.tar.gz https://get.helm.sh/helm-v2.15.1-linux-amd64.tar.gz
+tar xvf helm.tar.gz
+mv linux-amd64/helm /usr/local/bin/
+```
+Then we can set up the k8s cluster with the script
+```
+hack/kind-cluster-build.sh
+```
+Now we have our environment done!
+
+## Develop A HelloWorldChaos
+After preparing the development environment, let's develop a new type of chaos, HelloWorldChaos, which only prints a "hello world" message to log.
+
+As we know, the chaos is managed by the controller manager, so we should do something with the controller manager to add our HelloWorldChaos. 
+
+Check out the [main.go](https://github.com/pingcap/chaos-mesh/blob/master/cmd/controller-manager/main.go#L104) of controller manager we can see there are already some types of chaos:
+```golang
+	if err = (&controllers.PodChaosReconciler{
+		Client: mgr.GetClient(),
+		Log:    ctrl.Log.WithName("controllers").WithName("PodChaos"),
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "PodChaos")
+		os.Exit(1)
+	}
+
+	if err = (&controllers.NetworkChaosReconciler{
+		Client: mgr.GetClient(),
+		Log:    ctrl.Log.WithName("controllers").WithName("NetworkChaos"),
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "NetworkChaos")
+		os.Exit(1)
+	}
+
+	if err = (&controllers.IoChaosReconciler{
+		Client: mgr.GetClient(),
+		Log:    ctrl.Log.WithName("controllers").WithName("IoChaos"),
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "IoChaos")
+		os.Exit(1)
+	}
+```
+There are PodChaos, NetworkChaos and IoChaos. Let's add HelloWorldChaos:
+```golang
+	if err = (&controllers.HelloWorldChaosReconciler{
+		Client: mgr.GetClient(),
+		Log:    ctrl.Log.WithName("controllers").WithName("HelloWorldChaos"),
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "HelloWorldChaos")
+		os.Exit(1)
+	}
+```
+Of course this won't work since we didn't implement HelloWorldChaosReconciler yet, we should implement it in [controllers](https://github.com/pingcap/chaos-mesh/tree/master/controllers).
+
+Add a new file helloworldchaos_controller.go:
+```golang
+package controllers
+
+import (
+	"github.com/go-logr/logr"
```

Comment:
oh, why introducing another log lib, we have used many different log libs in PingCAP org now.

Reply:
> oh, why introducing another log lib, we have used many different log libs in PingCAP org now.

@cwen0 I agree with @siddontang, it's better to use uniform log lib.

Reply:
this log is unity with `controller-runtime`